File: v2/snippets/jar-library/TestClass.java
public class TestClass {
    public TestClass() {
    }

    // <Fields>
    public int publicValue;
    private int privateValue;
    public static int staticValue = 3;

    public TestClass(int publicValue, int privateValue) {
        this.publicValue = publicValue;
        this.privateValue = privateValue;
    }
    // </Fields>

    // <Methods>
    public static int multiplyByTwo(int a) {
        return 2 * a;
    }

    public int multiplyTwoNumbers(int a, int b) {
        return a * b;
    }
    // </Methods>

    // <Arrays>
    public String[] get1DArray() {
        return new String[]{"one", "two", "three", "four", "five"};
    }

    public String[][] get2DArray() {
        return new String[][]{{"S00", "S01"}, {"S10", "S11"}};
    }

    public double addArrayElementsAndMultiply(Double[] myArray, double myValue) {
        Double sum = 0.0;
        for (Double value : myArray) {
            sum += value;
        }
        return sum * myValue;
    }
    // </Arrays>

    // <Exceptions>
    public static int divideBy(int x, int y) {
        return divideBySecond(x, y);
    }

    public static int divideBySecond(int x, int y) {
        return divideByThird(x, y);
    }

    public static int divideByThird(int x, int y) {
        return x / y;
    }
    // </Exceptions>

    // <Cast>
    public static String castSampleMethod(Vehicle v) {
        return "castSampleMethod with Vehicle called";
    }

    public static String castSampleMethod(Car c) {
        return "castSampleMethod with Car called";
    }

    public static String castSampleMethod(Boat b) {
        return "castSampleMethod with Boat called";
    }
    // </Cast>

    // <Generics>
    public static <T> String genericSampleStaticMethod(T x, T y) {
        return x.toString() + " and " + y.toString();
    }

    public <T> String genericSampleMethod(T x, T y) {
        return x.toString() + " or " + y.toString();
    }

    public <T, K> String genericSampleMethodWithTwoTypes(T x) {
        return "genericSampleMethodWithTwoTypes invoked";
    }
    // </Generics>

    // <Enums>
    public enum Fruit {
        Apple,
        Banana,
        Orange,
        Mango
    }

    private static java.util.ArrayList<Fruit> Fruits = new java.util.ArrayList<Fruit>();

    public static String addFruitsToList(Fruit[] fruits) {
        Fruits.addAll(java.util.Arrays.asList(fruits));
        return String.format("%d fruits on the list", Fruits.size());
    }
    // </Enums>

    // <PassingNull>
    public static String passNull(Object arg) {
        if (arg == null) {
            return "Method called with null";
        } else {
            return "Method not called with null";
        }
    }

    public static String passNull2(int arg1, Object arg2) {
        if (arg2 == null) {
            return "Method2 called with null";
        } else {
            return "Method2 not called with null";
        }
    }
    // </PassingNull>

    // <ReturningNull>
    public static Object returnNull() {
        return null;
    }
    // </ReturningNull>

    // <Multithreading>
    private static java.util.concurrent.ConcurrentMap<Integer, Integer> cache = new java.util.concurrent.ConcurrentHashMap<>();

    public int addTwoNumbers(int x, int y) {
        System.out.println("Computing result in thread: " + Thread.currentThread().getId());
        try {
            Thread.sleep(1000); // Simulate computation delay
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        int result = x + y;
        System.out.println("Saving result in thread: " + Thread.currentThread().getId());
        try {
            Thread.sleep(50); // Simulate saving delay
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        cache.put((int) Thread.currentThread().getId(), result);
        System.out.println("Returning result in thread: " + Thread.currentThread().getId());
        return cache.get((int) Thread.currentThread().getId());
    }
    // </Multithreading>

    // <AsyncMethods>
    private static final java.util.concurrent.locks.ReentrantLock fileLock = new java.util.concurrent.locks.ReentrantLock();

public java.util.concurrent.CompletableFuture<Void> writeToFile(String fileName, String input) {
    return java.util.concurrent.CompletableFuture.runAsync(() -> {
        try {
            Thread.sleep(2000); // Simulate writing delay
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        fileLock.lock();
        try {
            java.nio.file.Files.write(java.nio.file.Paths.get(fileName), input.getBytes(), java.nio.file.StandardOpenOption.APPEND);
        } catch (java.io.IOException e) {
            throw new RuntimeException(e);
        } finally {
            fileLock.unlock();
        }
    });
}

    public int addThreeNumbers(int x, int y, int z) {
        try {
            Thread.sleep(2000); // Simulate computing delay
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return x+y+z;
    }
    // </AsyncMethods>

    // <Delegates>
    public static Integer useYourFunc(java.lang.reflect.Method yourFunc, Integer x, Integer y) throws Exception {
        return (Integer) yourFunc.invoke(null, x, y);
    }

    @FunctionalInterface
    public interface MathOperation {
        int operate(int x, int y);
    }

    public int useYourDelegate(MathOperation mathOperation, int x, int y) {
        return mathOperation.operate(x, y);
    }
    // </Delegates>


    // <Others>
    public Double multiplyTwoDoubles(Double a, Double b) {
        return a * b;
    }

    public double multiplyTwoDoublesPrimitive(double a, double b) {
        return a * b;
    }

    public Double addArrayElements(Double[] myArray) {
        Double sum = 0.0;
        for (Double value : myArray) {
            sum += value;
        }
        return sum;
    }

    public Double publicDoubleValue;

    public TestClass(Double publicDoubleValue) {
        this.publicDoubleValue = publicDoubleValue;
    }

    public static byte[] getByteArray() {
        byte[] test =  new byte[100000];
        java.util.Arrays.fill(test, (byte) 1);
        return test;
    }
    // </Others>

    // <Empty>
    // empty
    // </Empty>
}

File: v2/snippets/net-dll/TestClass.cs
using System.Collections.Generic;

namespace TestClass
{
	public class TestClass
	{
		public TestClass() { }

		// <Fields>
		public int PublicValue { get; set; }
		private int PrivateValue { get; }
		public static int StaticValue = 3;

		public TestClass(int publicValue, int privateValue)
		{
			PublicValue = publicValue;
			PrivateValue = privateValue;
		}
		// </Fields>

		// <Methods>
		public static int MultiplyByTwo(int a)
		{
			return 2 * a;
		}

		public int MultiplyTwoNumbers(int a, int b)
		{
			return a * b;
		}
		// </Methods>

		// <Arrays>
		public string[] Get1DArray()
		{
			return new string[] { "one", "two", "three", "four", "five" };
		}

        public string[,] Get2DArray()
        {
            return new string[,] { { "S00", "S01" }, { "S10", "S11" } };
        }

        public double AddArrayElementsAndMultiply(double[] myArray, double myValue)
		{
			double sum = 0;
			foreach (double element in myArray)
			{
				sum += element;
			}

			return sum * myValue;
		}
		// </Arrays>

		// <Exceptions>
		public static int DivideBy(int x, int y)
		{
			return DivideBySecond(x, y);
		}

		public static int DivideBySecond(int x, int y)
		{
			return DivideByThird(x, y);
		}

		public static int DivideByThird(int x, int y)
		{
			return x / y;
		}
		// </Exceptions>

		// <Cast>
		public static string CastSampleMethod(System.UInt32 value)
		{
			return "CastSampleMethod with System.UInt32 called";
		}

		public static string CastSampleMethod(System.Single value)
		{
			return "CastSampleMethod with System.Single called";
		}

		public static string CastSampleMethod(System.Double[] value)
		{
			return "CastSampleMethod with System.Double[] called";
		}
        // </Cast>

        // <Generics>
        public static string GenericSampleStaticMethod<T>(T x, T y)
        {
            return x + " and " + y;
        }

        public string GenericSampleMethod<T>(T x, T y)
        {
            return x + " or " + y;
        }

        public K GenericSampleMethodWithTwoTypes<T, K>(T x)
        {
            return default(K);
        }
        // </Generics>

        // <Enums>
		public enum Fruit
		{
			Apple,
			Banana,
			Orange,
			Mango
		}

		public static List<Fruit> Fruits = new List<Fruit>();

		public static string AddFruitsToList(Fruit[] fruits)
		{
            Fruits.AddRange(fruits);
			return string.Format("{0} fruits on the list", Fruits.Count);
		}
        // </Enums>

		// <Refs>
		public static void RefSampleMethod(ref int x)
		{
			x = x * 2;
		}

        public static void RefSampleMethod2(ref int x, ref double y, ref string z)
        {
            x = x * 2;
			y = y / 2;
			z = "Done";
        }
        // </Refs>

        // <Outs>
        public static void OutSampleMethod(out string outStr)
        {
			outStr = "String from OutSampleMethod";
        }
        // </Outs>

        // <Empty>
        // empty
        // </Empty>
    }
}
